src/java/org/apache/cassandra/repair/Scheduler.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.repair;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.concurrent.GuardedBy;
+
+import org.agrona.collections.LongArrayList;
+import org.apache.cassandra.utils.TimeUUID;
+import org.apache.cassandra.utils.concurrent.AsyncFuture;
+
+public interface Scheduler
+{
+    void schedule(TimeUUID sessionId, Executor taskExecutor, Task<?> tasks);
+
+    static Scheduler build(int concurrentValidations, SharedContext ctx)
+    {
+        return concurrentValidations <= 0
+               ? new NoopScheduler()
+               : new LimitedConcurrentScheduler(concurrentValidations, ctx);
+    }
+
+    final class NoopScheduler implements Scheduler
+    {
+        @Override
+        public void schedule(TimeUUID sessionId, Executor taskExecutor, Task<?> tasks)
+        {
+            taskExecutor.execute(tasks);
+        }
+    }
+
+    final class LimitedConcurrentScheduler implements Scheduler
+    {
+        private final int concurrentValidations;
+        private final SharedContext ctx;
+        @GuardedBy("this")
+        private int inflight = 0;
+        @GuardedBy("this")
+        private final Map<TimeUUID, Group> groups = new HashMap<>();
+
+        LimitedConcurrentScheduler(int concurrentValidations, SharedContext ctx)
+        {
+            this.concurrentValidations = concurrentValidations;
+            this.ctx = ctx;
+        }
+
+        @Override
+        public synchronized void schedule(TimeUUID sessionId, Executor taskExecutor, Task<?> tasks)
```

Comment:
the usage is a single task, so all sessions will have the same `score`... so a simpler data structure could be used... this class makes the most sense when we have way to compare tasks

Reply:
I have changed it to a single queue of task-executor pairs.